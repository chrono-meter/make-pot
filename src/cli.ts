#!/usr/bin/env node

import process from "node:process";
import * as yargs from "yargs";
import { hideBin } from "yargs/helpers";
import { getArgs } from "./cli/getArgs.js";
import { getJsonArgs } from "./cli/getJsonArgs.js";
import makeJsonCommand from "./jsonCommand.js";
import makepotCommand from "./potCommand.js";
import type { Args, MakeJsonArgs } from "./types.js";

/** Main execution */
// Get the selected command
const r = yargs
	.default(hideBin(process.argv))
	.options({
		makejson: {
			describe: "Make JSON file",
			type: "boolean",
			default: false,
		},
	})
	.parseSync() as { makejson: boolean };

// Execute the command
if (!r.makejson) {
	makepotCommand(getArgs() as Args);
} else {
	makeJsonCommand(getJsonArgs() as MakeJsonArgs);
}
