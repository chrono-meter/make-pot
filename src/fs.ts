import type { Args } from './types'
import path from 'path'
import fs from 'fs/promises'
import { generatePotHeader } from './utils'

export async function writePotFile (args: Args, fileContent: string): Promise<void> {
  // the path to the .pot file
  const potFilePath = path.join(args.destination, `${args.slug}.pot`)

  await fs.writeFile(potFilePath, generatePotHeader(args) + fileContent)
}
