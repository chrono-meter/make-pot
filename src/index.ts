#!/usr/bin/env node
import { makePot } from './makePot'

import { getArgs } from './cliArgs'

import yargs from 'yargs'

import { name, version } from '../package.json'

/** Main execution */
const args = getArgs()
if (Object.keys(args).length > 0) {
	console.log(name + ' version: ' + version)
	const timeStart = new Date()
	makePot(args)
		.then(() => {
			const timeEnd = new Date()
			console.log(
				`ðŸš€ Translation Pot file created in ${timeEnd.getTime() - timeStart.getTime()}ms`
			)
		})
		.catch((error) => {
			console.error(error)
		})
} else {
	yargs.showHelp()
}
