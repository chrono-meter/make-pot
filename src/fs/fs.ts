import fs, { writeFileSync } from "node:fs";
import { readFile } from "node:fs/promises";
import path from "node:path";
import type { Args } from "../types";

/**
 * Ensures that a folder exists at the specified path.
 *
 * @param {string | undefined} folderPath - The path of the folder to ensure existence for.
 * @return {string} - The path of the folder, or '.' if folderPath is undefined.
 */
function ensureFolderExists(folderPath: string | undefined): string {
	if (folderPath === undefined) {
		return ".";
	}
	try {
		// Check if the folder exists
		fs.accessSync(
			path.resolve(folderPath),
			fs.constants.R_OK | fs.constants.W_OK,
		);
	} catch (error) {
		if ((error as NodeJS.ErrnoException).code === "ENOENT") {
			// The Folder does not exist, so create it
			fs.mkdirSync(folderPath, { recursive: true });
			console.log(`Folder created: ${folderPath}`);
			return folderPath;
		}
	}
	return folderPath;
}

/**
 * Gets the charset of the .pot file
 *
 * @param charset the charset of the .pot file
 * @return the charset of the .pot file
 */
export function getCharset(charset: string | undefined): BufferEncoding {
	if (!charset) {
		return "latin1";
	}
	// we need to check if the charset is valid otherwise we return latin1 that is a common alias for ISO-8859-1 and the default charset for pot files
	switch (charset.toLowerCase()) {
		case "utf-8":
			return "utf-8";
		default:
			return "latin1";
	}
}

export function getEncodingCharset(charset: string | undefined): string {
	if (!charset) {
		return "iso-8859-1";
	}
	// we need to check if the charset is valid otherwise we return utf-8 that is a common alias for ISO-8859-1 and the default charset for pot files
	switch (charset.toLowerCase()) {
		case "utf-8":
			return "utf-8";
		default:
			return "iso-8859-1";
	}
}

/**
 * The output path for the pot file.
 * @param outpath - the output path for the pot/json files
 * @return {string} - the output path
 */
export function getOutputPath(outpath?: string): string {
	return path.join(process.cwd(), outpath ?? "languages");
}

/**
 * The output path for the pot file.
 * @param args - the command line arguments
 */
function getOutputFilePath(args: Args): string {
	const { out, headers, options } = args.paths;
	let i18nFolder = out ?? headers?.domainPath ?? "languages";

	// Remove leading and trailing slashes
	i18nFolder = i18nFolder.replace(/^\/+|\/+$/g, "");

	const extension = options?.json ? "json" : "pot";

	return path.join(process.cwd(), i18nFolder, `${args.slug}.${extension}`);
}

/**
 * Writes the .pot file to disk
 *
 * @param fileContent the content of the .pot file
 * @param args the command line arguments passed to the program
 */
export function writeFile(fileContent: string, args: Args): void {
	const dest = getOutputFilePath(args);

	if (ensureFolderExists(path.dirname(dest))) {
		// get the encoding charset
		const encodingCharset = getCharset(args.options?.charset);
		console.log(`File created at ${dest}`);

		// write the file
		const potBuffer = Buffer.from(fileContent);
		writeFileSync(dest, potBuffer.toString(encodingCharset), {
			encoding: encodingCharset,
		});
	} else {
		console.log(`Folder ${dest} does not exist and cannot be created`);
	}
}

/**
 * The async version of fs.readFile method
 * @param path the path of the file to read
 * @return the content of the file as a string
 */
export async function readFileAsync(path: string): Promise<string> {
	return readFile(path, "utf-8");
}
