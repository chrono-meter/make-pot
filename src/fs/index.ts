import * as path from 'path'
import fs from 'node:fs'

/**
 * Ensures that a folder exists at the specified path.
 *
 * @param {string | undefined} folderPath - The path of the folder to ensure existence for.
 * @return {string} - The path of the folder, or '.' if folderPath is undefined.
 */
function ensureFolderExists(folderPath: string | undefined): string {
	if (folderPath === undefined) {
		return '.'
	}
	try {
		// Check if the folder exists
		fs.accessSync(
			path.resolve(folderPath),
			fs.constants.R_OK | fs.constants.W_OK
		)
	} catch (error) {
		if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
			// The Folder does not exist, so create it
			fs.mkdirSync(folderPath, { recursive: true })
			console.log(`Folder created: ${folderPath}`)
			return folderPath
		}
	}
	return folderPath
}

/**
 * Writes the .pot file to disk
 *
 * @param fileContent the content of the .pot file
 * @param dest the path of the .pot file to write
 */
export function writeFile(fileContent: string, dest: string) {
	if (ensureFolderExists(path.dirname(dest))) {
		fs.writeFileSync(dest, fileContent)
	}
}

export async function readFileAsync(path: string): Promise<string> {
	return fs.promises.readFile(path, 'utf-8')
}
