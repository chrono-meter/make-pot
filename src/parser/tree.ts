import Parser, { type SyntaxNode } from 'tree-sitter'
import { type TranslationStrings } from '../types'
import { i18nFunctions } from '../const'

import { GetTextComment, GetTextTranslation } from 'gettext-parser'
import { getParser } from '../fs/glob'
import { stripTranslationMarkup } from '../utils'

/**
 * Collect comments from the AST node and its preceding siblings.
 *
 * @param {SyntaxNode} node - The AST node.
 * @return {string[]} An array of collected comments.
 */
function collectComments(node: SyntaxNode): string | undefined {
	let currentNode = node
	let depth = 0

	// Check the node's preceding siblings for comments
	while (currentNode && depth < 6) {
		if (
			currentNode?.previousSibling?.type === 'comment' &&
			currentNode?.previousSibling?.text
				.toLowerCase()
				.includes('translators')
		) {
			return stripTranslationMarkup(currentNode?.previousSibling.text)
		}
		depth++
		currentNode = currentNode.parent as SyntaxNode
	}
}

/**
 * Parses the source code using the specified language parser and extracts the strings from the file.
 *
 * @param {string} sourceCode - The source code to be parsed.
 * @param {string} filepath - The path to the file being parsed.
 * @return {TranslationStrings[]} An array of translation strings.
 */
export function doTree(
	sourceCode: string,
	filepath: string
): TranslationStrings {
	// set up the parser
	const parser = new Parser()
	parser.setLanguage(getParser(filepath))

	// parse the file
	const tree = parser.parse(sourceCode)

	const gettextTranslations: TranslationStrings = {}
	const typeToMatch =
		filepath.split('.').pop()?.toLowerCase() !== 'php'
			? 'call_expression'
			: 'function_call_expression'

	/**
	 * Traverse the tree ðŸŒ³
	 *
	 * @param {SyntaxNode} node The node to traverse through
	 */
	function traverse(node: SyntaxNode): void {
		// Walk the tree
		if (node?.children.length)
			for (const child of node.children) {
				traverse(child)
			}

		// Check if the node matches
		if (node?.type === typeToMatch) {
			// The function name is the first child
			const functionName = node.firstChild?.text ?? null
			if (
				functionName === null ||
				!Object.keys(i18nFunctions).includes(functionName)
			) {
				return
			}

			// The arguments are the last child
			const argsNode = node.lastChild
			if (
				argsNode === null ||
				argsNode.childCount === 0 ||
				argsNode.type !== 'arguments'
			) {
				return
			}

			// Get the whole gettext translation string
			// eslint-disable-next-line @typescript-eslint/no-unused-vars
			const [_fn, raw] = node.children
			const translation: Partial<GetTextTranslation> = {}

			const translationKeys =
				i18nFunctions[functionName as keyof typeof i18nFunctions]

			const children = raw.children.slice(1, -1)
			let translationKeyIndex = 0

			// Get the translation from the arguments (the quoted strings)
			for (const child of children) {
				let node = child
				let nodeValue = node.text

				// unwrap the argument node, which is used in PHP.
				if (child.type === 'argument') {
					if (child.children.length === 0) continue
					node = child.children[0]
				}

				if (node?.type === ',') {
					// skip the comma between arguments
					continue
				} else if (
					node?.type === 'string' ||
					node.type === 'encapsed_string'
				) {
					// unquote the strings
					nodeValue = nodeValue.slice(1, -1)
				}

				// the translation key (eg. msgid)
				const currentKey = translationKeys[translationKeyIndex]

				// the value of that key
				translation[currentKey as keyof typeof translation] = nodeValue

				// increment the index of the translation key
				translationKeyIndex += 1
			}

			// Get the translation data
			const gettext: GetTextTranslation = {
				msgctxt: translation.msgctxt ?? '',
				msgid: translation.msgid ?? '',
				msgid_plural: translation.msgid_plural ?? '',
				msgstr: translation.msgid_plural ? ['', ''] : [''],
				comments: {
					translator: collectComments(node) ?? '',
					reference: `${filepath}:${node.startPosition.row + 1}`,
				} as GetTextComment,
			}

			gettextTranslations[gettext.msgctxt ?? ''] = {
				...(gettextTranslations[gettext.msgctxt ?? ''] || {}),
				[gettext.msgid]: gettext,
			}
		}
	}

	traverse(tree.rootNode)

	// Return both matches and entries
	return gettextTranslations
}
