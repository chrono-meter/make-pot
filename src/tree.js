#!/usr/bin/env node
import Parser from 'tree-sitter'
import Ts from 'tree-sitter-typescript'
import Js from 'tree-sitter-javascript'
import Php from 'tree-sitter-php'
import path from 'path'
import { glob } from 'glob'
import { readFile } from 'fs/promises'

/**
 * Extract strings from a file ðŸŒ³
 *
 * @param {Tree} node
 * @param {string} lang
 * @param {string} filename
 * @return {{}}
 */
function extractStrings (node, lang, filename) {
  const matches = []
  const typeToMatch = lang !== Php ? 'call_expression' : 'function_call_expression'

  /**
   * Traverse the tree ðŸŒ³
   *
   * @param {Tree} node The node to traverse through
   * @param {Tree|null} previousNode The previous node in the tree
   */
  function traverse (node, previousNode = null) {
    if (node.type === typeToMatch) {
      const functionNameNode = node.firstChild
      let meta = {}
      const functionName = functionNameNode !== null ? functionNameNode.text : null

      if (functionName?.match(/^(_n|_x|_nx|__|_e)$/) !== null) {
        const argsNode = node.lastChild
        const [fnName, ...msgid] = node.children.map(argNode => argNode.text)

        // Extract the metadata
        if (argsNode !== null && argsNode.childCount > 0 && argsNode.type === 'arguments') {
          // Extract the msgid
          const firstArgNode = argsNode.firstChild
          const msgidNode = firstArgNode.firstChild
          const msgid = msgidNode?.text

          // Extract the msgctxt if it exists and unquote it
          const msgctxtNode = firstArgNode.nextNamedSibling
          let msgctxt
          if (msgctxtNode !== null && msgctxtNode.type === 'string') {
            // unquote the string
            msgctxt = msgctxtNode.text.slice(1, -1)
          }

          let comments
          // Find the preceding comment node at the same level as the translation function call.
          let nodeToCheck = previousNode
          while (nodeToCheck !== null && nodeToCheck.type !== 'comment') {
            nodeToCheck = nodeToCheck.previousSibling
          }
          // If the node is a comment, extract the comment text
          if (nodeToCheck !== null && nodeToCheck.type === 'comment') {
            const commentText = nodeToCheck.text.trim()
            if (commentText.includes('translators:') !== false) {
              comments = commentText
            }
          }

          const position = node.startPosition
          meta = {
            msgid,
            msgctxt,
            comments,
            reference: `#: ${filename}:${position.row + 1}:${position.column + 1}` // Adjusted to be 1-indexed
          }
        }

        matches.push({ fnName, ...meta, msgid })
      }
    }

    if (node.children.length > 0) {
      for (const child of node.children) {
        // Make sure to pass the current node as previousNode to the next level of recursion.
        traverse(child, node)
      }
    }
  }

  traverse(node)

  // Return both matches and entries
  return matches
}

/**
 * Parse a file and extract strings
 *
 * @param {object} args
 * @param {string} args.filepath - Path to the file to parse
 * @param {string} args.language - Language of the file to parse
 * @return {{}}
 */
async function parseFile (args) {
  const sourceCode = await readFile(path.resolve(args.filepath), 'utf8')

  if (args.language === 'Text') {
    return { text: sourceCode }
  } else if (args.language === 'Json') {
    return JSON.parse(sourceCode)
  }

  const parser = new Parser()
  parser.setLanguage(args.language)
  const tree = parser.parse(sourceCode)

  return extractStrings(tree.rootNode, args.language, args.filepath) // Assuming extractStrings is synchronous
}

/**
 * Return the parser based on the file extension
 * @param file - Path to the file
 * @return {*|{}|string} - Parser
 */
function getParser (file) {
  const ext = file.split('.').pop()
  switch (ext) {
    case 'ts':
      return Ts.typescript
    case 'tsx':
      return Ts.tsx
    case 'js':
    case 'jsx':
    case 'mjs':
    case 'cjs':
      return Js
    case 'php':
      return Php
    case 'json':
      return 'Json'
    default:
      return 'Text'
  }
}

async function getFiles (src = 'sourcedir/**') {
  return await glob(src, { nodir: true, cwd: process.cwd() })
}

async function run () {
  const files = await glob('sourcedir/**', { nodir: true, cwd: process.cwd() })
  const parsePromises = files.map(async (file) => {
    console.log('Searching for strings in ' + file)
    const parsed = await parseFile({ filepath: file, language: getParser(file) })
    console.log(parsed)
    return parsed
  })

  const results = await Promise.all(parsePromises)
  console.log(results.flat())
}
const startTime = Date.now()
await run()
console.log('ðŸŽ‰ Done in', Date.now() - startTime, 'ms.')
