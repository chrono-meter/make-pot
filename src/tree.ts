#!/usr/bin/env node
import Parser, { type SyntaxNode, Tree } from 'tree-sitter'
// @ts-expect-error
import Js from 'tree-sitter-javascript'
// @ts-expect-error
import Php from 'tree-sitter-php'
// @ts-expect-error
import Ts from 'tree-sitter-typescript'
import path from 'path'
import { readFile } from 'fs/promises'
import { type TranslationString } from './types'
import { i18nFunctions } from './const'

/**
 * Extract strings from a file ðŸŒ³
 *
 * @param {Tree} node
 * @param {string} lang
 * @param {string} filename
 * @return {{}}
 */
export function extractStrings (node: SyntaxNode, lang: Parser | null, filename: string): TranslationString[] {
  const matches: TranslationString[] = []
  const typeToMatch = lang !== Php ? 'call_expression' : 'function_call_expression'

  /**
   * Traverse the tree ðŸŒ³
   *
   * @param {SyntaxNode} node The node to traverse through
   * @param {SyntaxNode|null} previousNode The previous node in the tree
   */
  function traverse (node: SyntaxNode, previousNode: SyntaxNode | null = null): void {
    if (node.type === typeToMatch) {
      const functionNameNode = node.firstChild
      const functionName = functionNameNode !== null ? functionNameNode.text : null

      if (functionName && Object.keys(i18nFunctions).includes(functionName)) {
        const argsNode = node.lastChild

        // Extract the metadata
        if (argsNode !== null && argsNode.childCount > 0 && argsNode.type === 'arguments') {
          // Extract the msgid
          const firstArgNode = argsNode.firstChild!
          const [fn, raw] = node.children.map(argNode => argNode.text.trim())

          let comments
          // Find the preceding comment node at the same level as the translation function call.
          let nodeToCheck = previousNode
          while (nodeToCheck !== null && nodeToCheck.type !== 'comment') {
            nodeToCheck = nodeToCheck.previousSibling
          }
          // If the node is a comment, extract the comment text
          if (nodeToCheck !== null && nodeToCheck.type === 'comment') {
            const commentText = nodeToCheck.text.trim()
            if (commentText.includes('translators:')) {
              comments = commentText
            }
          }

          // the function used for translation
          const type = i18nFunctions[fn as keyof typeof i18nFunctions] || 'text_domain'
          // The position of the translation in the file
          const position = node.startPosition
          // extract the values from the raw translation - ATM the content of the parsed raw is a sting like ('translation', 'td') and maybe with formatting characters like \n or \t
          const parsedRaw = raw
            .slice(1, -1) // remove enclosing parentheses
            .split(',') // split on comma
            .map(str => str.trim().slice(1, -1)) // remove enclosing quotes

          matches.push({
            type,
            raw: parsedRaw,
            msgid: parsedRaw[0],
            comments,
            reference: `#: ${filename}:${position.row + 1}:${position.column + 1}` // Adjusted to be 1-indexed
          } as TranslationString)
        }
      }
    }

    if (node.children.length > 0) {
      for (const child of node.children) {
        // Make sure to pass the current node as previousNode to the next level of recursion.
        traverse(child, node)
      }
    }
  }

  traverse(node)

  // Return both matches and entries
  return matches
}

/**
 * Parse a file and extract strings asynchronously
 *
 * @param {object} args
 * @param {string} args.filepath - Path to the file to parse
 * @param {Parser|null} args.language - Language of the file to parse
 * @return {Promise<TranslationString[]>}
 */
export async function parseFile (args: { filepath: string, language: Parser | null }): Promise<TranslationString[]> {
  // log the filepath
  console.log('Parsing ', args.filepath)

  // read the file
  const sourceCode = await readFile(path.resolve(args.filepath), 'utf8')

  if (args.language === null) {
    return [{
      type: '',
      raw: [],
      msgid: [sourceCode],
      reference: '#: ' + args.filepath + '  } '
    }]
  }

  // set up the parser
  const parser = new Parser()
  parser.setLanguage(args.language)

  // parse the file
  const tree = parser.parse(sourceCode) // Assuming parse is an async operation

  // extract the strings from the file and return them
  return extractStrings(tree.rootNode, args.language, args.filepath)
}

/**
 * Return the parser based on the file extension
 * @param file - Path to the file
 * @return {*|{}|string} - Parser
 */
export function getParser (file: string): null | Parser {
  const ext = file.split('.').pop()
  switch (ext) {
    case 'ts':
      return Ts.typescript
    case 'tsx':
      return Ts.tsx
    case 'js':
    case 'jsx':
    case 'mjs':
    case 'cjs':
      return Js
    case 'php':
      return Php
    default:
      return null
  }
}
