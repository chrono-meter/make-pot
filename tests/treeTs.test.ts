import { doTree } from '../src/tree'
import path from 'path'

import fs from 'fs'

describe('doTree', () => {
	it('Should parse TSX file and extract strings', async () => {
		const fileContent = fs.readFileSync(
			path.join(process.cwd(), 'tests/fixtures/block/SvgControls.tsx'),
			'utf8'
		)
		const fileParsed = doTree(fileContent, 'SvgControls.tsx')
		console.log(fileContent.slice(0, 500), fileParsed)
		expect(fileParsed).toMatchSnapshot()
	})
})
