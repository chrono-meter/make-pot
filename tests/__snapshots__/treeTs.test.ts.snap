// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doTree tsx Should parse TSX file and extract strings 1`] = `
{
  "": {
    "Edit Link": {
      "comments": {
        "reference": "SvgControls.tsx:93",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Edit Link",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Replace SVG": {
      "comments": {
        "reference": "SvgControls.tsx:107",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Replace SVG",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Unlink": {
      "comments": {
        "reference": "SvgControls.tsx:99",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Unlink",
      "msgid_plural": "",
      "msgstr": [],
    },
  },
}
`;
